driver/src/main/java/org/neo4j/driver/internal/InternalSession.java

```diff
@@ -121,6 +121,27 @@ public TypeSystem typeSystem()
         return InternalTypeSystem.TYPE_SYSTEM;
     }
 
+    @Override
+    protected void finalize() throws Throwable
+    {
+        if( isOpen )
+        {
+            // Using stderr here to avoid calling into a logging system that may not exist when this finalizer runs.
+            System.err.println( "WARNING: Neo4j Session object leaked, please ensure that your application calls the `close` method on Sessions before disposing of the objects." );
+            if( currentTransaction != null )
+            {
+                // If there is an open transaction, rather than deal with the complexity of cleaning that up inside the GC finalizer,
+                // force the connction to be closed, rather than returned to the connection pool.
+                // This can be removed once we have a good way in Bolt to reset a session, then we won't need to close it like this.
+                connection.forceClose();
+            }
+            else
+            {
+                connection.close();
+            }
+        }
+    }
```

Comment:
Yeah, having a `final Log` type field would be better. The fields won't be nulled out until after the finalizer has completed. If the `Log` type has a finalizer method, then I'm not sure if you can rely on any call order between the two, though. In Stormpot I just let the objects leak. Whatever the case, there is definitely (assuming a bug-free Java driver) a bug in user-code if you ever find the need to put out that log line.